amethyst_assets/src/loader.rs

```diff
@@ -314,42 +325,67 @@ pub fn load_asset_future<A, F, N, S>(context: Arc<A::Context>,
                                      name: N,
                                      storage: S,
                                      thread_pool: &ThreadPool)
-                                     -> AssetFuture<A, AssetError<A::Error, F::Error, S::Error>>
+                                     -> AssetFuture<A, F::Error, S::Error>
     where A: Asset + Send + 'static,
           A::Context: Context + Send + Sync + 'static,
           A::Error: Send + 'static,
+          A::Result: Send + 'static,
           F: Format<Data = A::Data> + Send + 'static,
           F::Error: Send + 'static,
           N: Into<String>,
           S: Store + Send + Sync + 'static
 {
     let name = name.into();
 
-    let closure = move || load_asset::<A, F, _, S>(&*context, &format, name, &storage);
+    let store_id = storage.store_id();
+    let spec = AssetSpec::new(name.clone(), F::extension(), store_id);
 
-    AssetFuture::spawn(thread_pool, closure)
+    match context.retrieve(&spec) {
+        Some(asset) => AssetFuture::new(spec, asset),
+        None => {
+            AssetFuture::spawn(spec.clone(), thread_pool, move || load_asset_inner::<A, F, S>(&context, &format, &spec, &storage))
+        }
+    }
 }
 
 /// Like `reload_asset`, but loads the asset on a worker thread and returns
 /// an `AssetFuture` immediately.
-pub fn reload_asset_future<A, F, N, S>
-    (context: Arc<A::Context>,
-     format: F,
-     name: N,
-     storage: S,
-     thread_pool: &ThreadPool)
-     -> AssetFuture<A, AssetError<A::Error, F::Error, S::Error>>
+pub fn reload_asset_future<A, F, N, S>(context: Arc<A::Context>,
```

Comment:
If you upgrade `Format` and `Store` to use `IntoFuture` than it will be good idea